Http multi threading

bazsi opened the pull request:
This branch implements multithreaded http destination, that uses the specified number of worker threads to send messages to an upstream HTTP server.

This should drastically improve performance of HTTP based transports, especially if combined with #2289, #2287, #2286

kira-syslogng commented:
Build FAILURE

bazsi commented:
This one depends on #2279, as I've removed the patch from the series that works around a problem, and without that, this one crashes.

My expectation is that if #2279 goes in, this should turn green, sans any unrelated problems (which now is riemann).

I just wanted to publish this, for anyone to see what I've been working on in the recent weeks.

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

Kokan commented:
@bazsi could you please resolve the conflict ?

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

bazsi commented:
This now depends on #2302 that it also implicitly contains. If that goes in, this could be next.

I am doing another round of manual testing if it survived all the rebase rounds :)

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

lbudai commented:
@kira-syslogng do perftest

kira-syslogng commented:
Build FAILURE

lbudai commented:
@pzoleex : is this the same issue that you've found internally?

lbudai commented:
@bazsi : additional information: the issue is not introduced by your PR, we think it is triggered by the stats related changes (it is possible that it is not _caused_ by that stats PR but somehow it triggered something...), so your PR still mergeable and doesn't require any action from you

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

furiel commented on modules/http/http.c:
```diff
@@ -202,20 +202,40 @@ _add_message_to_batch(HTTPDestinationWorker *self, LogMessage *msg)
+                 log_pipe_location_tag(&owner->super.super.super.super));
       retval = WORKER_INSERT_RESULT_ERROR;
       break;
     default:
```
Cosmetic, but I would prefer default statement to be the last case.

bazsi replied:
the default falls through to the 5XX case that's why it was in that order. But I've added a followup patch, that among other things makes "default" the last one.

furiel commented on modules/http/http.c:
```diff
@@ -24,569 +24,656 @@
+      msg_error("Unknown HTTP response code",
+                evt_tag_str("url", owner->url),
+                evt_tag_int("status_code", http_code),
+                log_pipe_location_tag(&owner->super.super.super.super));
```
We should set an error code in the default case too. Otherwise memory garbage is returned.

bazsi replied:
the default was falling through to the 5xx case, so a default was set. but my followup patch makes that absolutely clear.

furiel commented on lib/logthrdestdrv.h:
```diff
@@ -63,6 +65,7 @@ struct _LogThreadedDestWorker
   gint batch_size;
   gint rewound_batch_size;
   gint retries_counter;
+  gint seq_num;
```
I think this should remain gint32: step_sequence_number_atomic returns with gint32.

bazsi replied:
good catch.

furiel commented on lib/logthrdestdrv.c:
```diff
@@ -888,9 +931,9 @@ log_threaded_dest_driver_init_instance(LogThreadedDestDriver *self, GlobalConfig
   self->lock = g_mutex_new();
-  self->worker.instance.started_up = g_cond_new();
+  log_threaded_dest_worker_init_instance(&self->worker.instance, self);
+  _init_worker_compat_layer(&self->worker.instance);
```
This line should be moved before init_instance (->init_watches)

furiel replied:
My bad, confused with start_watches. Deleting my comment.

furiel commented on lib/logthrdestdrv.c:
```diff
@@ -804,6 +850,17 @@ log_threaded_dest_driver_set_max_retries(LogDriver *s, gint max_retries)
 }
 
+LogThreadedDestWorker *
+_lookup_worker(LogThreadedDestDriver *self, LogMessage *msg)
```
I would call select_worker or pick_worker or similar. As we have an array of workers, maybe we need a real lookup method later.

bazsi replied:
Agree

bazsi replied:
I would change this name in a followup branch, not this one, as it already received some feedback and I have stacks of followup work on top of this.

But please call me on this should I not address that in the load balancing patches.

Thanks.

furiel commented on lib/logthrdestdrv.c:
```diff
@@ -886,7 +944,16 @@ log_threaded_dest_driver_init_method(LogPipe *s)
+  gboolean startup_success = TRUE;
+  for (gint i = 0; startup_success && i < self->num_workers; i++)
+    {
+      startup_success &= _start_worker_thread(self);
```
The deinit method needs similar code for the those threads that are successfully initialized. self->workers_started might help with that.

bazsi replied:
workers are automatically terminated at exit, so I think they are cleared up just fine.

main_loop_create_worker_thread() gets a function callback that is invoked when a thread needs to exit due to a reload/stop being processed. In that case _request_worker_exit() gets called, which sets a variable and posts an iv_event.

Then the mainloop mechanism will wait for all worker threads to shut down and that's when deinit() gets called. By that time workers exited.

I think that addresses the issue, do you agree?

furiel replied:
I was thinking that the individual workers also have init deinit mechanism, but that's not the case as only the "master" has init-deinit, hence was my confusion. I agree. Marking as resolved.

furiel commented:
@bazsi about the last patch: http: split HTTPDestinationWorker, can you explain the changes or a general idea behind it?

I expect that should be a straigthforward patch that just follows up the changes in the logthrdestdrv driver. But  as methods are moved around in the same patch, it makes difficult to follow the diff.

Should we look out for something special about that patch?

bazsi commented on modules/http/http.c:
```diff
@@ -202,20 +202,40 @@ _add_message_to_batch(HTTPDestinationWorker *self, LogMessage *msg)
     {
-    case 4:
-      retval = WORKER_INSERT_RESULT_DROP;
+    case 1:
```
good catch, I have set this to return an ERROR and not NOT_CONNECTED. I think this should be rather rare, and if our retry mechanism does not solve it, we shouldn't repeat the same batch.

I have fixed it in a followup patch.

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

bazsi commented:
Responding to https://github.com/balabit/syslog-ng/pull/2291#issuecomment-425629002, @furiel 

Unfortunately at one point I squashed the http refactors into a single patch, but it is should be nothing but extracting HTTPDestinationWorker from the Driver, changing the signatures of all functions and moving code so that code for each of the classes are in one block. I'll see if I can resplit the patch. The riemann destination was similarly refactored, but I did that in 3 steps, but it's not going to be simpler than that.

bazsi commented:
I have now extracted the biggest chunk of code moves around in that patch, making it much easier to review. The patch that moves the code around was actually redone now and squashed out of the refactor, so I am 100% certain that no other changes than code moves are in that patch.

I hope this helps with the review.

furiel commented:
Thank you @bazsi for the resplit! It helped a lot with the review.

gaborznagy commented:
Fixes #2276 !

bazsi commented:
On Fri, Oct 5, 2018, 14:27 Gábor Nagy <notifications@github.com> wrote:

> *@gaborznagy* commented on this pull request.
> ------------------------------
>
> In lib/logthrdestdrv.c
> <https://github.com/balabit/syslog-ng/pull/2291#discussion_r223025110>:
>
> >
>    _schedule_restart(self);
>    iv_main();
>
> +  log_threaded_dest_worker_flush(self);
> +  log_queue_rewind_backlog_all(self->queue);
>
> It's a minor thing, not affecting behaviour:
> Shouldn't we highlight that rewinding after flush should happen only if
> the flush failed and the queue has remaining messages?
> We could use the log_threaded_dest_worker_flush return value.
>

rewinding an empty backlog will do nothing. I don't see much value in
checking that.

Actually I am rather thinking about a flush-on-exit() parameter that would
control whether:

1) we flush batched data but only that
2) we flush the entire queue
3) we don't flush at all.

The default for persistent queues would be 1), non persistent queues would
be 2). This would allow the user to solve cases where the destination is
slow, and we dont want to wait the flush of the entire queue on each reload.


------------------------------
>
> In lib/logthrdestdrv.c
> <https://github.com/balabit/syslog-ng/pull/2291#discussion_r223039070>:
>
> >    self->startup_finished = TRUE;
> +  self->startup_failure |= thread_failure;
>
> Just out of curiosity: why do we need to use bitwise OR when setting
> failure?
> It looks like if workers can be started multiple times and we don't want a
> failure to be masked out.
>
We want to fail if any of the workers failed.



—
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/2291#pullrequestreview-162059009>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AArldo7z0BFP91c1o6hO6Ht8gp0F3eyKks5uh6SPgaJpZM4Wt7qb>
> .
>

gaborznagy commented on lib/logthrdestdrv.c:
```diff
@@ -555,70 +559,292 @@ _init_watches(LogThreadedDestDriver *self)
-  g_mutex_lock(self->lock);
+  g_mutex_lock(self->owner->lock);
   self->startup_finished = TRUE;
+  self->startup_failure |= thread_failure;
```
> We want to fail if any of the workers failed.
My point was that `startup_failure` is a Worker attribute and not the Driver's, therefore I don't think it can be evaluated multiple times, hence .
But it was just a question, it caught my attention, it looked interesting. :)

lbudai commented on lib/logthrdestdrv.h:
```diff
@@ -116,7 +119,13 @@ struct _LogThreadedDestDriver
+   * static for a single insert() invocation, whereas this might be
+   * increased in parallel by the multiple threads. */
+
+  gint32 shared_seq_num;
```
Shouldn't we use a 64 bit atomic counter (`atomic-gsize`) for `shared_seq_num`?
It will overflow after ~3hours when the message rate is 400k/sec.
Or it's not a problem as it is used with timestamp?  and (currently :) ) we cannot do 2^32-1 msg/sec...

bazsi replied:
It's only meant to be used for ordering and is explicitly sized as 32 bits:

quoting rfc5424:

7.3.1.  sequenceId

   The "sequenceId" parameter tracks the sequence in which the
   originator submits messages to the syslog transport for sending.  It
   is an integer that MUST be set to 1 when the syslog function is
   started and MUST be increased with every message up to a maximum
   value of 2147483647.  If that value is reached, the next message MUST
   be sent with a sequenceId of 1.
